FINERACT-700: upgrade Gradle from 2.10 to 5.2.1

vorburger opened the pull request:
https://issues.apache.org/jira/browse/FINERACT-700

vorburger committed:
upgrade Gradle version, and deal with impacts (FINERACT-700)

https://discuss.gradle.org/t/could-not-get-unknown-property-classesdir-for-main-classes-of-type-org-gradle-api-internal-tasks-defaultsourcesetoutput/30570

https://discuss.gradle.org/t/gradle-4-replacing-deprecated-test-testclassesdir/22822/2

https://discuss.gradle.org/t/could-not-find-method-leftshift-for-arguments-on-task-of-type-org-gradle-api-defaulttask/30614

vorburger committed:
upgrade spring-boot-gradle-plugin, and deal with impacts (FINERACT-700)

because the old one previously used caused an UnsupportedOperationException
with the new Gradle, see https://issues.apache.org/jira/browse/FINERACT-700

The spring-jms and activemq-broker versions are set to what they were on before.

vorburger committed:
upgrade (switch) OpenJPA Gradle plugin (FINERACT-700)

from at.schmutterer.oss.gradle:gradle-openjpa
to com.radcortez.gradle:openjpa-gradle-plugin

due to https://github.com/schmutterer/gradle-openjpa/issues/8

vorburger commented:
would anyone watching this repo have any interest in and be able to help review (test?) this?

vishwasbabu commented on fineract-provider/build.gradle:
```diff
@@ -364,7 +363,7 @@ task integrationTest(type:Test){
         tomcatRunWar.daemon = true
         tomcatRunWar.execute()
```
With the removal of TaskInternal.execute() method in Gradle 5.x , this task (integrationTest) would fail

michaelbukachi replied:
Also all flyway tasks fail since they are using `TaskInternal.execute()`

michaelbukachi replied:
I tinkered a bit using the specified `gradle 5.2.1` and discovered that the specified flyway plugin `3.0` is not compatible with the gradle version due to the above mentioned error. I discovered that in flyway `3.1` they added a new feature where one can override `FlywayTask` so implementing the flyway tasks using the below format would get rid of the `TaskInternal.execute()` error:
```

task migrateTenantListDB(type: org.flywaydb.gradle.task.FlywayMigrateTask) {
    description="Migrates a Tenant List DB. Optionally can pass dbName. Defaults to 'mifosplatform-tenants' (Example: -PdbName=someDBname)"
    def filePath = "filesystem:$projectDir" + '/src/main/resources/sql/migrations/list_db'
    def tenantsDbName = 'mifosplatform-tenants';
    if (rootProject.hasProperty("dbName")) {
        tenantsDbName = rootProject.getProperty("dbName")
    }
    extension = new org.flywaydb.gradle.FlywayExtension()
    extension.url= "jdbc:mysql:thin://localhost:3306/$tenantsDbName"
    extension.driver = "org.drizzle.jdbc.DrizzleDriver"
    extension.user = 'root'
    extension.password = '.....'
    extension.locations= [filePath]
}
```

However, in gradle `5.0` the property `sourceSet.output.classesDir` was removed in favour of `sourceSet.output.classesDirs`. You'll notice that in the internal implementation of Flyway 3.1 [here](https://github.com/flyway/flyway/blob/f7fa24196789e2cbcc5ffffb409b5eca2c629846/flyway-gradle-plugin/src/main/groovy/org/flywaydb/gradle/task/AbstractFlywayTask.groovy) the propety `classesDir` is still being used:
```
def runTask() {
        if (isJavaProject()) {
            def classLoader = Thread.currentThread().getContextClassLoader()
            project.sourceSets.each {
                def classesUrl = it.output.classesDir.toURI().toURL() <----------
                logger.debug("Adding directory to Classpath: " + classesUrl)
                classLoader.addURL(classesUrl)
        ...............
}
```
Flyway `5.0.0` was corrected to support both `classesDir` and `classesDirs`, however, it doesn't seem to be compatible with `drizzle-jdbc`; even the latest version `1.4`. I tried the default `mysql-connector` but it always threw a bunch of errors during migration. Some are show below:
```
DB: Unknown table 'acc_gl_account' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'acc_gl_closure' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'acc_gl_journal_entry' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'acc_product_mapping' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'c_configuration' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_appuser' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_appuser_role' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_calendar' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_calendar_instance' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_charge' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_client' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_client_identifier' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_code' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_code_value' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_currency' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_deposit_account' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_deposit_account_transaction' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_document' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_fund' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_group' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_group_level' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_group_client' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_guarantor' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_charge' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_arrears_aging' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_collateral' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_officer_assignment_history' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_repayment_schedule' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_loan_transaction' (SQL State: 42S02 - Error Code: 1051)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: No data - zero rows fetched, selected, or processed (SQL State: 02000 - Error Code: 1329)
DB: Unknown table 'm_template' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_templatemappers' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'm_template_m_templatemappers' (SQL State: 42S02 - Error Code: 1051)
DB: Unknown table 'sms_messages_outbound' (SQL State: 42S02 - Error Code: 1051)
DB: Data truncated for column 'last_time_password_updated' at row 1 (SQL State: 01000 - Error Code: 1265)
DB: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '--Journal entries also refer to Client transactions

```
 To remove those errors changes will have to be made the `.sql` files.

muarachmann replied:
@michaelbukachi following this and it is a quite interesting discovery for me. I have pulled this in and switched to a new branch. So i can work on this . Have been following all along and i see some changes have to be made to the sql file. 

What changes are to be made. I am a newbie and this depends on an issue i am trying to solve for integrating spotBugs - here https://issues.apache.org/jira/browse/FINERACT-702 which depends on a latest version of gradle.

I would love to help in removing unwanted stuff or update the sql files accordingly thanks

michaelbukachi replied:
These are the changes which need to made:

- Updating of the sql files to support latest mysql/mariadb versions. Note. This can be quite tedious due to the fact there are around 360+ `.sql` files to go through.
- Once that is done, you'll need to change the drizzle driver to something else. You'll have to choose between mysql and mariadb jdbc drivers. Note. Drizzle is used all over the place in different files so you might have to do a search
- Next, you'll have to update the flyway plugin
- Finally, you can update gradle to the latest version
- Remember to run integration tests in the end to make sure everything is good.

muarachmann replied:
Thanks alot. I am on it. will ping you if i have an error

vorburger replied:
> pulled this in and switched to a new branch. So i can work on this

@muarachmann if it's possible to raise small incremental PRs, e.g. first doing some SQL migrations, if they are backwards compatible, it makes it easier for other people to review smaller changes and merge them gradually.

> choose between mysql and mariadb jdbc drivers

IMHO we should use the [MariaDB JDBC Driver](https://mariadb.com/kb/en/library/about-mariadb-connector-j/), not the (Oracle..) MySQL driver. I know less about Drizzle, but thought that it was abandoned; note drizzle.org is dead & last org.drizzle.jdbc:drizzle-jdbc released in Aug, 2016, so I would ignore that. Again as I said above, it may be best to do this in separate gradual incremental Pull Requests.

awasum replied:
@vorburger Is the MariaDB driver license compatible with the Apache License? The ASF prohibits the use of non compliant (category X) libraries on Apache projects. Maybe there is the work around for this? @myrle-krantz , @vishwasbabu

michaelbukachi replied:
@awasum seems MariaDB driver is under the LGPL license which [seems](https://www-us.apache.org/legal/resolved.html#category-x) not to be compatible with ASF

ivange94 replied:
@michaelbukachi I spoke with @muarachmann over the phone and he's trying to update .sql files to be compliant with mysql 8. I'm not really clear as to how upgrading gradle to 5.2.1 will require a mysql upgrade. I understand that the drizzle plugin is not compatible, but shouldn't we just be replacing it with a compatible plugin that still maintains the current mysql syntax?

> Updating of the sql files to support latest mysql/mariadb versions. Note. This can be quite tedious due to the fact there are around 360+ .sql files to go through.

I feel the above should have been in a separate issue IMO.

michaelbukachi replied:
@ivange94 It's because the flyway plugin is affected. There current plugin isn't compatible with gradle 5. Upgrading the plugin would require upgrading the driver.

cnangeh replied:
> @michaelbukachi I spoke with @muarachmann over the phone and he's trying to update .sql files to be compliant with mysql 8. I'm not really clear as to how upgrading gradle to 5.2.1 will require a mysql upgrade. I understand that the drizzle plugin is not compatible, but shouldn't we just be replacing it with a compatible plugin that still maintains the current mysql syntax?
> 
> > Updating of the sql files to support latest mysql/mariadb versions. Note. This can be quite tedious due to the fact there are around 360+ .sql files to go through.
> 
> I feel the above should have been in a separate issue IMO.

I think he said he didn't find a compatible drizzle-jdbc plugin. But @muarachmann can still try to do some research on getting a compatible drizzle plugin

vorburger replied:
Resolving this old and now outdated conversation, because it's no longer relevant for this PR.

vorburger commented:
Ah yes, didn't realize that, good point worth checking out more... (I
wonder how other Java projects at Apache.org use MariaDB, or even MySQL.)

PS: Not sure I totally understand how the Gradle upgrade necessarily
requires changing the currently used JDBC driver.. if that could somehow be
split off to be dealt separately and later, that could be cool.

>

ivange94 commented:
@vorburger I have pulled your branch and tried building and running the application.

**gradle build** completes successfully but with this weird output in the console. [No metadata was found for 262 classes; they have been enhanced as persistence-aware. If you intended for any of these classes to be persistence-capable, then this means that OpenJPA could not find any metadata for those classes. This can happen if the directory containing your metadata is not in your CLASSPATH, or if your metadata files are not named properly.](https://pastebin.com/VTJkpgJz)

**gradle tomcatRunWar** fails with https://pastebin.com/jePJfcet

Output with --stacktrace enabled https://pastebin.com/bL6BpxtG

vorburger commented:
I've just rebased this on the current `develop`, and resolved the conflict it had in `fineract-provider/dependencies.gradle`; my main motivation is to have include #528, so that the Travis build also runs integration tests. I expect that with that it will now fail to build on TravisCI (before the rebsae it used to pass and had "All checks have passed"), which is a Good Thing, because then we can work on getting it to pass - somehow.

https://issues.apache.org/jira/browse/FINERACT-761 tracks how to resolve the Drizzle/Flyway mess...

michaelbukachi commented:
Awesome! I'm following this closely. This is related to pull request #550. Which has stalled due it becoming a legal matter.

vorburger committed:
upgrade gradle-tomcat-plugin from 1.0 to 2.5 (NOK)

vorburger commented:
I've made further progress on this, and now almost works, except `./gradlew integrationTest` fails to start Tomact... one possible solution for this could be [FINERACT-764](https://issues.apache.org/jira/browse/FINERACT-764) - unless anyone else has a smarter idea how to fix it?

vorburger committed:
gradle.properties with -Xmx4g and disabling daemon

because the (new, upgraded) Tomcat Gradle plugin needs more memory
(and this seems to make the build a bit faster, may be less is fine),
and it does not work with the Gradle daemon; this fixes:

> Task :tomcatRunWar
Failed to initialize end point associated with ProtocolHandler ["http-bio-8080"]
Failed to initialize connector [Connector[HTTP/1.1-8080]]
Failed to initialize end point associated with ProtocolHandler ["ajp-bio-8009"]
Failed to initialize connector [Connector[AJP/1.3-8009]]
Failed to initialize end point associated with ProtocolHandler ["http-bio-8443"]
Failed to initialize connector [Connector[HTTP/1.1-8443]]

which appears on the 2nd (only!) run of `./gradlew tomcatRunWAR`
(because the Tomcat embedded in the previous daemon-ized Gradle did not stop).

vorburger committed:
fix gradle createDB/dropDB/setBlankPassword

which broke during the Gradle upgrade in FINERACT-700

do not fix Flyway related migrateTenantDB & Co, as those
are proposed to be removed in FINERACT-765 anyway.  (If really
needed, some of my other WIP branches have the required changes
to make the Flyway related tasks work.)

vorburger commented:
I suggest that we ignore the entire Flyway related discussion (above, in previous inline review discussions), I think I actually got it work in the end now, but all that was more of a "distraction" re. the main point of this effort (the Gradle upgrade) - because we (probably) will just remove Flyway from Gradle anyway, see #613.

But what we do still need to do here is to get `./gradlew integrationTest`to pass on this branch/PR - it currently does not, because Tocmcat does not fully start (it starts but seems to ignore Fineract WAR completely), see `build/reports/tests/integrationTest/classes/org.apache.fineract.integrationtests.AccountNumberPreferencesTest.html`.

The interesting thing is that `./gradlew tomcatRunWAR` does work though! I've not yet understood what the difference / root cause of this is, and would very much welcome input from anyone who is able to dig more into this... anyone able to help and step up here?

PS: In the mean time I started exploring just completely ditching Tomcat and using only Spring Boot, but I have concluded that may be too big a step in one go - if at all possible, I think it would be much better & clearer to just wrap up this PR with a more narrow focused upgrade of Gradle and the minimal required adjustments, only - and then only separately after follow-up re. running integration tests with Spring Boot instead of Tomcat (FINERACT-764).

michaelbukachi commented:
@vorburger  Let me look into this. I've successfully made integration test run locally.

vorburger commented:
I've just rebased this on latest `develop`, and resolved conflicts with recently merged changes.

As before, `./gradlew integrationTest` still fails. @michaelbukachi have you had any luck debugging this?

michaelbukachi commented:
@vorburger Not yet. I'm trying to run `./gradlew tomcatRun` to make sure the server is starting smoothly, but I'm getting a ` Could not get unknown property 'classesDir' for main classes of type org.gradle.api.internal.tasks.DefaultSourceSetOutput`. **classesDir** was deprecated in 4 and removed 5. I'm trying to find out which plugin is causing it so that I can update it.

michaelbukachi commented:
@vorburger It seems the tomcat plugin being used, `org.gradle.api.plugins:gradle-tomcat-plugin:1.0` is incompatible with **gradle 5**. Changing it to `com.bmuschko:gradle-tomcat-plugin:2.5` and also changing `apply plugin: 'tomcat` to `apply plugin: 'com.bmuschko.tomcat'` removes the error.

michaelbukachi commented:
@vorburger Are we still using the drizzle driver?

vorburger commented:
> @vorburger Not yet. I'm trying to run `./gradlew tomcatRun` to make sure the server is starting smoothly, but I'm getting

how weird that I didn't hit this...

> @vorburger It seems the tomcat plugin being used, `org.gradle.api.plugins:gradle-tomcat-plugin:1.0` is incompatible with **gradle 5**. Changing it to `com.bmuschko:gradle-tomcat-plugin:2.5` and also changing `apply plugin: 'tomcat` to `apply plugin: 'com.bmuschko.tomcat'` removes the error.

Duh! Yes, I had actually done EXACTLY what you describe in a previous incarnation of this PR, but somehow in rebases and conflict resolutions skipped over and lost it... but now found it again (e.g. as 0590ea11aeb721f97e0d21153b5928674dc08b00) and cherry-picked it back onto this PR. Perhaps you could like to pull this and confirm that for you it works with this as it stands now?

We'll see on Travis, but for me locally it still does not work, but due to what I think is a new and (finger crossed) other problem, watch [FINERACT-778](https://issues.apache.org/jira/browse/FINERACT-778) ... if you don't hit that one, it could perhaps be due to OpenJDK vs. Oracle.

> @vorburger Are we still using the drizzle driver?

yeah, see #613 and [FINERACT-765](https://issues.apache.org/jira/browse/FINERACT-765) and note _this does not affect e.g. " ./gradlew createDB -PdbName=mifosplatform-tenants", that will still work (and is really quite useful), because createDB has nothing to do with Flyway._

lingocoder commented:
> how weird that I didn't hit this...

Heads-up @vorburger, About an hour or so ago (*probably before you wrote your above comment*) I pulled down [*your java11 branch*](http://bit.ly/VrbrgrJ11) to see what I could make of [*the apache#609 failure*](http://bit.ly/Trvs609J11) you mentioned in our earlier conversation. 

The very first thing I always do immediately after pulling down a new branch is I run *`gradle check`* to make sure all is well. This is what I got on your *`java11`* branch:

```
FAILURE: Build failed with an exception.
...
* What went wrong:
A problem occurred evaluating root project 'fineract-provider'.
> Could not get unknown property 'classesDir' for main classes of type org.gradle.api.internal.tasks.DefaultSourceSetOutput.
...
BUILD FAILED in 2m 19s
```

A grep on *`classesDir`* turned up this:

```
$ grep -n -R 'classesDir' fineract-provider
fineract-provider/build.gradle:75:    files = fileTree(sourceSets.main.output.classesDir).matching {
fineract-provider/build.gradle:329:sourceSets.main.output.resourcesDir = sourceSets.main.output.classesDir
fineract-provider/build.gradle:330:sourceSets.test.output.resourcesDir = sourceSets.test.output.classesDir
fineract-provider/build.gradle:369:    testClassesDir = project.sourceSets.integrationTest.output.classesDir

```

lingocoder commented:
> how weird that I didn't hit this...

P.S. @vorburger , [*The apache#609 CI failure*](http://bit.ly/Trvs609J11) is most likely due to the last change to [*fineract-provider/gradle/wrapper/gradle-wrapper.properties*](http://bit.ly/GrdlWrpr210) being four+ years ago. Version 2.10 of the wrapper needs to be updated to at least 5.2.x I believe.

vorburger commented:
> We'll see on Travis, but for me locally it still does not work, but due to what I think is a new and (finger crossed) other problem, watch [FINERACT-778](https://issues.apache.org/jira/browse/FINERACT-778) ... if you don't hit that one, it could perhaps be due to OpenJDK vs. Oracle.

I actually cannot reproduce FINERACT-778 anymore, but `./gradlew integreationTest` fail both on TravisCI and locally for me with the same `ConnectException` - basically, it seems, in ITs Tomcat isn't starting Fineract? The interesting thing is that `./gradlew tomcatRunWAR` does work .. and I thought that did, roughly, the same.. I've not yet understood what the difference / root cause of this is, and would very much welcome input from anyone who is able to dig more into this... anyone able to help and step up here?

@lingocoder the points you are raised above are all related to Java 11 only, which I'd like to exclude from this PR. FYI I used to want to to Gradle upgrade, Java 11, Spring boot ITs all in one, but have come to the conclusion that it will be better if we go about this step by step. I'll therefore copy/paste and reply to the points you are raised above re. Java 11 over in #609, and ignore them here.

vorburger committed:
fix integrationTest (=> runWar) which broke in Gradle upgrade

inspired by (but not exactly like) the respective section on
https://github.com/bmuschko/gradle-tomcat-plugin#faq

vorburger commented:
>  basically, it seems, in ITs Tomcat isn't starting Fineract? The interesting thing is that ./gradlew tomcatRunWAR does work .. and I thought that did, roughly, the same.. 

Oh! I just figured it out. Basically, as was actually pretty obvious just from staring at the console (as always, once you see it...), `./gradlew integrationTest` indeed just no longer started Tomcat - that `doFirst` thing to launch `tomcatRunWar.daemon = true` just seems to not work anymore when upgrading  Gradle from 2.10 to 5.2.1 (and there is no error message about that at all, shame on you, Gradle). With the latest additional commit just pushed to this PR now, it is starting to look much better for me locally - now let's see if all ITs pass on Travis CI with the upgraded Gradle!

vorburger commented:
> All checks have passed (1 successful check)
> continuous-integration/travis-ci/pr — The Travis CI build passed 

Yay! Party time. :cake: :birthday: :tada: 

This finalizes this PR from my side. Anyone reading this, please code review and shout if anything NOK.

Committers (notably e.g. @vishwasbabu @awasum @Anh3h) will any of you merge this?

vorburger committed:
Merge branch 'develop' into bump-gradle-wrapper

vorburger commented:
@conradsp would you perhaps like to review/test/merge this?

awasum commented:
Closing to reopen again to trigger Travis

awasum commented:
Gradle tasks: clean, build are running fine.

Tasks, tomcatRunWar and :tomcatRunWarAsDaemonForIntegrationTest' fail with: 
```
Execution failed for task ':tomcatRunWar'.
> java/util/logging/LogRecord
```
On Googling, I got this: https://stackoverflow.com/questions/23963049/classcircularityerror-java-util-logging-logrecord-running-gradle-webapp-with-ja

@vorburger  You actually already fixed the problem in the past as seen the the build.gradle file and on SO link above. So I dont know whats going wrong now. Maybe am missing something.

awasum commented:
```
> Task :bootRun FAILED
Error: Could not find or load main class org.apache.fineract.ServerWithMariaDB4jApplication

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':bootRun'.
> Process 'command '/usr/lib/jvm/java-8-oracle/bin/java'' finished with non-zero exit value 1
```

vorburger commented:
@awasum it used to pass, then we merged #607, then I merged `develop` into this PR _"just to make sure"_ , and boum - somehow that change broke this again.. I'm just having a look to see if I can figure out what's going on...

cnangeh reviewed (commented):
@vorburger I just tested this and this task fails, **compileIntegrationTestJava**. I think the is an issue with the log configuration.

vorburger committed:
fix ClassCircularityError: java/util/logging/LogRecord (see FINERACT-700 for details)

vorburger commented:
The last commit I just added to this PR will, hopefully, fix the problem above; see last comment in [FINERACT-700](https://issues.apache.org/jira/browse/FINERACT-700) for some details of what happened and how I fixed it. - Travis is currently running, and I'm going to bed now; cross fingers it will pass... :smiley: 

@awasum @Anh3h thanks for looking into this, testing, and therefore helping to get this merged!

awasum commented:
Everything looks fine on my local branch. Thanks @vorburger

awasum reviewed (approved):
LGTM

cnangeh reviewed (approved):
Looks good to me.

vorburger commented:
Great! I'll "self merge" this PR now.
